Enterprise-CMCS/managed-care-review
# MR-3117: Attempt to update unmodifiable data

## Summary

Following up on @JasonLin0991 's suspicion that this bug was actually caused by failing to disable the Continue and Back buttons in time (he was right) I worked to debounce...something.  Initially, I debounced in the UI, but then thought that it would be nice to have a way to debounce resolver calls more generally (the front-end useDebouncer will live on in git history if we ever want to resurrect it).

With the useDebouncedMutation hook, we can debounce any mutation simply by wrapping it in the hook.

~Our best guess as to what's causing this bug is that in some cases, when the server call after `Continue` takes longer than usual, and the user navigates back to already-edited page before the call returns, and _then_ tries to continue again, we're trying to update with stale cache data.~

~Ultimately, this bug is possible because we don't handle potential concurrent editing.  If two people are simultaneously working on one submission, one shouldn't constantly see stale data warnings.~

~After chatting with @macrael, we decided that at some point we'll have to deal with the concurrent editing problem, but that's far out of scope for this ticket and bug (which I can't even reproduce, after much trying).  So we're going to update the error message to give the user some guidance about how to proceed, and we'll put [a ticket](https://qmacbis.atlassian.net/browse/MR-3353) in the backlog for concurrent editing.~

#### QA Guidance

You should not be able to trigger extra update calls, no matter how fast you click Continue.

#### Test cases covered

Once I finally figured out why our spies weren't working (exporting via index.ts can cause jest spies to not work), I still had a tough time getting the kind of test I really wanted, which was to click many times and see only one update.  My best guess is that the way Jest triggers clicks doesn't allow multiple fast clicks (I tried loops, doubleClick events, and several other strategies.)  The best I could do was confirm that the debouncer was called when update was called.
